File: code/CodeSphere-front/src/share/type/problem.ts
export interface UserProblem extends ProblemFake {
  status: 'solved' | 'attempted' | 'unsolved';
  bookmarked: boolean;
  category: string;
}

export interface ProblemFake {
  id: number;
  title: string;
  difficulty: '쉬움' | '보통' | '어려움';
  successRate: number;
  tags: string[];
  submissions: number;
  likes: number;
  estimatedTime: string;
}

// ---- 위에껀 사용 안됨 -----
export interface Problem {
  real_pid: string;
  title: string;
  tag: string[];
  submit_count: number;
  user_result: 'PASS' | 'FAIL' | 'NONE';
  correct_rate: number;
  level: number;
}

export interface ProblemDetail {
  real_pid: string;
  title: string;
  body: string;
  example_io: ExampleIo[];
  level: number;
  tag: string[];
  input: string;
  output: string;
  problem_constraint: string;
}

export interface ExampleIo {
  input: string;
  output: string;
}

File: code/CodeSphere-back/app/schemas/problem.py
from typing import List, Literal, Optional
from pydantic import BaseModel, ConfigDict

class ExampleIO(BaseModel):
    input: str
    output: str

class ProblemOut(BaseModel):
    real_pid: int
    title: str
    body: str
    example_io: list[ExampleIO]
    level: int
    tag: list[str]
    input: Optional[str] = None
    output: Optional[str] = None
    problem_constraint: Optional[str] = None

    class Config:
        from_attributes = True

class ProblemListItem(BaseModel):
    real_pid: int
    title: str
    level: int
    tag: list[str]
    submit_count: int
    correct_rate: float  # 0.0 ~ 1.0
    user_result: Literal["PASS", "FAIL", "NONE"] = "NONE"

    model_config = ConfigDict(from_attributes=True)

class ProblemListResponse(BaseModel):
    total: int
    data: List[ProblemListItem]